tools/scripts/bumpenvs.yaml

```diff
@@ -0,0 +1,13 @@
+gcp_env: pedl-environments-06318c7
+tf1_cpu_hashed: determinedai/environments:py-3.6.9-pytorch-1.4-tf-1.15-cpu-06318c7
+tf1_cpu_versioned: determinedai/environments:py-3.6.9-pytorch-1.4-tf-1.15-cpu-0.5.0
+tf1_gpu_hashed: determinedai/environments:cuda-10.0-pytorch-1.4-tf-1.15-gpu-06318c7
+tf1_gpu_versioned: determinedai/environments:cuda-10.0-pytorch-1.4-tf-1.15-gpu-0.5.0
```

Comment:
Is updating these part of this PR (didn't see them bumped anywhere)

Reply:
Yes, the first two commits do not update the task environments.  The third commit is a taskenv bump, which was created by me just applying this script.

Reply:
Sorry I mean the versioned taks envs specifically (`determinedai/environments:cuda-10.0-pytorch-1.4-tf-1.15-gpu-0.5.0`).

Reply:
Damn!  I didn't realized t hat we used these forms of the tags in the sphinx docs.  I think it may be back to the drawing board.

@dzhu looks like bumpenvs.py is not going to cut it.  My next best idea is to switch from a config file to a state file, where the state file keeps pairs of (old tag, new tag) and the update operation replaces all the old tag strings with the new tag strings.  I don't like making it stateful but I can't think of any other way to make bumpenvs.py work with snippets like these (without getting way too complicated): 

```.. code:: yaml

   environment:
     image:
       gpu: "determinedai/environments:cuda-10.1-pytorch-1.4-tf-2.2-gpu-0.5.0"
       cpu: "determinedai/environments:py-3.6.9-pytorch-1.4-tf-2.2-cpu-0.5.0"
```

Reply:
Also @aaron276h, what do you think?

Reply:
sounds reasonable to me, that is more or less how bump version works.

Reply:
Perhaps you considered these already, but some possibilities that came to my mind are (1) allowing multiple BUMPENVS lines in a row to translate into multiple output lines and (2) allowing a BUMPENVS line to contain `\n` to translate into multiple lines. I see the complications, but those still look a bit better to me than making the whole thing stateful and based on replacements in existing text.

Reply:
Ok, I think supporting multiple consecutive BUMPENVS lines is... palatable.  I could imagine something like this:

```rst
.. BUMPENVS: .. code:: yaml
.. BUMPENVS:
.. BUMPENVS:    environment:
.. BUMPENVS:      image:
.. BUMPENVS:        gpu: "${tf2_gpu_versioned}"
.. BUMPENVS:        cpu: "${tf2_cpu_versioned}"
.. code:: yaml

   environment:
     image:
       gpu: "determinedai/environments:cuda-10.1-pytorch-1.4-tf-2.2-gpu-0.5.0"
       cpu: "determinedai/environments:py-3.6.9-pytorch-1.4-tf-2.2-cpu-0.5.0"
```

I'm a little torn on if the increase in complexity is worth the statelessness.  I *think* I care more about the statelessness.  But playing devil's advocate, in defense of the stateful solution, literally all of the tags we are replacing are fundamentally unique identifiers... it's not nearly as dangerous as `bumpversion` (which is why `bumpversion` takes a whitelist of files and `bumpenvs.py` is ok to just `git grep` the whole repository).

Additionally, for the stateless solution, I've been thinking we really need to add a linting check as PR check to make sure that `bumpenvs.py` wouldn't make any changes if it were run.  I guess that might be seen as a pro for the stateless solution as "more robust" or a con for the stateless solution as "more overhead".

@aaron276h, @dzhu what is your guys' take here?

Reply:
Hmm, it occurs to me that the vagaries of reST in particular are going to make autoformatting, when it comes into play, interact poorly with that scheme: `.. <stuff>` is only a comment if the stuff _happens_ to not line up with a valid directive, so best practice is to always have `..` on a line by itself for comments, which `rstfmt` enforces, which really shouldn't change just for this. There could be multiple `BUMPENVS` lines in a comment block, but then they have to be indented, which means `bumpenvs` would have to gain the ability to selectively ignore leading whitespace to make the output right.

One other way this could work is using the `literalinclude` directive to move the `BUMPENVS` lines into a separate file:

`custom-env.txt`:
```rest
.. literalinclude:: envs.yaml
   :language: yaml
   :start-after: cpu:
```

`envs.yaml`:
```yaml
# BUMPENVS: environment:
# BUMPENVS:   image:
# BUMPENVS:     gpu: "${tf2_gpu_versioned}"
# BUMPENVS:     cpu: "${tf2_cpu_versioned}"
environment:
  image:
    gpu: "determinedai/environments:cuda-10.1-pytorch-1.4-tf-2.2-gpu-0.5.0"
    cpu: "determinedai/environments:py-3.6.9-pytorch-1.4-tf-2.2-cpu-0.5.0"
```

`literalinclude` could also specify `:lines: 2,4,6,8` to allow `bumpenvs` as it stands now to work without requiring the multi-line capability, though writing out the lines is a bit more annoying.

----

For the other spot with the environment names inline in a paragraph, we could use [substitutions](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#substitutions) to include the values in one place and use them elsewhere, which works around this.

----

So that all can work, but, having thought about how the unique nature of these identifiers means the update can essentially be one `sed` command, I'm now overall leaning toward biting the bullet and making it stateful to avoid going through any of these contortions and keep the update logic simple. I'm not very opposed to anything else that's come up, though.

Reply:
I personally prefer the stateful approach as it seem simpler, and I do not thing that maintaining this state correctly will be hard (famous last words).

Reply:
Stateful approach is ready for review now.